Demo slicing pipeline final

Huongg opened the pull request:
## Description

This PR is for QA-ing purpose.

I've merged all the latest changes from 3 different PRs here (#1980 , #1996, #2003)


## QA notes
This PR should cover the following tasks

**_As a Kedro user, I want to be able to select multiple nodes within the flowchart so that I can slice the pipeline I am currently viewing_**
- [x] Notification on First Node Click: When I click on the first node, a notification should appear. This behavior should not occur when clicking on a pipeline (is it referred to as "pipeline" or "modular pipeline"?).
- [x]  Second Node Click with Shift: If I hold the Shift key and click on a second node, the metadata should close immediately, and the notification should disappear.
- [x]  when selecting a node from bottom to top, the order should be updated according to --from-node=bottom-node-id --to-node=top-node-id

**_As a Kedro user, I want to preview which nodes are included in my selection for slicing so that I can understand what my slice will be before creating a new slice_**

- [x] Chart Highlighting: The two nodes selected with Shift held should be highlighted with a darker blue background, and any nodes in between should be highlighted with a lighter blue background.
- [x]  Nodes List Highlighting: The nodes list on the left-hand side should also highlight all the nodes included in the newly selected pipeline.
- [x]  Call to Action: A button at the bottom should display the number of nodes selected and include a "Slice" button.
- [x]  Outside Click: Clicking outside the selection will un-select all the nodes, and go back to the default/previous page. This includes clicking on any other nodes outside the sliced pipeline

**_As a Kedro user, I want to view my sliced pipeline and its corresponding run command so that I can run this slice in the CLI for my Kedro project. I also want to be able to reset my sliced pipeline view so that I can go back to viewing the full Kedro pipeline_**

- [x]  Filtered Nodes View: A new view should display all the filtered nodes.
- [x]  Updated Node List: The node list should reflect all the newly filtered nodes.
- [x]  CLI Command Display: A command line at the bottom should show how to run the filtered nodes in the CLI.
- [x]  Close the sliced view: Click the exit button should take the user back to viewing the full kedro pipeline


## Checklist

- [ ] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

stephkaiser commented:
hello! first round of QA comments from me (will do the final round when i'm back from PTO):

1. Notification padding on right side needs to be fixed. Also need to remove the ' ' 
<img width="397" alt="Screenshot 2024-08-13 at 11 16 32" src="https://github.com/user-attachments/assets/64736797-3c36-4c72-b227-44641f969855">

2. Padding for the slice action bar also needs to be fixed, the padding should be including the hover state, as you can see below its currently really tight between the copy button and the slice button and run command:
<img width="500" alt="Screenshot 2024-08-13 at 11 17 29" src="https://github.com/user-attachments/assets/99b28833-13a8-4b84-b709-786c336fbf26">

3. when a node is selected inside a pipeline group and in the list its collapsed, there is no blue indication line at the parent pipeline level - this was part of the highlighting inconsistencies issue (https://github.com/kedro-org/kedro-viz/pull/2004)

> selecting a child node and then closing the pipeline group doesnt bring the blue selection line to the parent level, can we also include this change? This would be useful in slicing when the contents of a pipeline are selected but not expanded
<img width="862" alt="Screenshot 2024-08-13 at 11 19 23" src="https://github.com/user-attachments/assets/ed6a525a-4420-474d-9c70-21a8752bce4f">

4. the blue indication line for selected nodes in light UI should be accent-blue-300 (a lighter shade of blue)
<img width="127" alt="Screenshot 2024-08-13 at 11 23 34" src="https://github.com/user-attachments/assets/3eb1a71c-3b4c-45cc-ad94-ff34e758f78c">

5. the gradient when the run command is really long and user can scroll it still has the same issue :( I also noticed a small gap:
<img width="66" alt="Screenshot 2024-08-13 at 11 27 25" src="https://github.com/user-attachments/assets/c38d9cda-314b-4c47-be5c-3c408f9f3017">

6. Reset slice button in light UI should have white colour text instead of black: (on hover it looks correct)
<img width="164" alt="Screenshot 2024-08-13 at 11 27 49" src="https://github.com/user-attachments/assets/62efe3cd-523e-4c0d-8273-6e7790babf68">

7. After slicing, the nodes are still selected in the list panel even though they are not currently selected in my slice view:
<img width="852" alt="Screenshot 2024-08-13 at 11 28 32" src="https://github.com/user-attachments/assets/522441b8-2d75-485a-ad94-646148ba9626">

8. Hard to describe this one but overall the animation of the notification appearing and the slice action bar appearing feels a little strange, maybe theres too much 'movement'.

Thank you for all this great work!!

ravi-kumar-pilla commented:
Hi @stephkaiser , I agree on this 
> Hard to describe this one but overall the animation of the notification appearing and the slice action bar appearing feels a little strange, maybe theres too much 'movement'.

I made a suggestion [here](https://github.com/kedro-org/kedro-viz/pull/1996#issuecomment-2276195517) for something subtle, may be in future, once you are back and have some time, please have a look. Thank you

Huongg commented:
hey @stephkaiser thank you for you review. I believe I've addressed all your points. 

- [ ] Notification padding on right side needs to be fixed. Also need to remove the ' '

- [ ] Padding for the slice action bar also needs to be fixed, the padding should be including the hover state, as you can see below its currently really tight between the copy button and the slice button and run command:
<img width="1710" alt="Screenshot 2024-08-16 at 10 39 26" src="https://github.com/user-attachments/assets/c695ec8e-c6a4-4269-b22e-b7bb2a361759">


- [ ] when a node is selected inside a pipeline group and in the list its collapsed, there is no blue indication line at the parent pipeline level - this was part of the highlighting inconsistencies issue (https://github.com/kedro-org/kedro-viz/pull/2004)
<img width="1710" alt="Screenshot 2024-08-16 at 10 39 26" src="https://github.com/user-attachments/assets/d80ce54f-1cfd-493a-8574-5d5903bf0bdc">

- [ ] selecting a child node and then closing the pipeline group doesnt bring the blue selection line to the parent level, can we also include this change? This would be useful in slicing when the contents of a pipeline are selected but not expanded


_**Is this one the same as the point above? If so then its addressed now**_

- [ ] the blue indication line for selected nodes in light UI should be accent-blue-300 (a lighter shade of blue)

- [ ] the gradient when the run command is really long and user can scroll it still has the same issue :( I also noticed a small gap:


**_I've now used this combination for the shadow. Let me know what you think_**
```
Dark theme:
 --sliced-pipeline-action-bar-shadow-start: #{colors.$black-500};
  --sliced-pipeline-action-bar-shadow-end: #{colors.$black-900};
Light theme
--sliced-pipeline-action-bar-shadow-start: #{colors.$white-900};
  --sliced-pipeline-action-bar-shadow-end: #{colors.$white-200};
```
<img width="1072" alt="Screenshot 2024-08-16 at 10 39 52" src="https://github.com/user-attachments/assets/5436bec8-c17a-402c-ac51-ccacd5d0f75a">

<img width="1015" alt="Screenshot 2024-08-16 at 10 39 59" src="https://github.com/user-attachments/assets/463adcc1-6ec6-4e50-bdbb-9a6ede086387">


- [ ] Reset slice button in light UI should have white colour text instead of black: (on hover it looks correct)

- [ ] After slicing, the nodes are still selected in the list panel even though they are not currently selected in my slice view:

- [ ] Hard to describe this one but overall the animation of the notification appearing and the slice action bar appearing feels a little strange, maybe theres too much 'movement'.


**_So I've implemented this new animation, inspired by @ravi-kumar-pilla suggestion. Let me know what you think?_**

- notification fades in at the start
- action bar slides up from the bottom
- action bar slides left-right to position in the middle of flowchart depends on whether node list or meta data panel opens or not


https://github.com/user-attachments/assets/518f2052-7444-459b-be07-94754cf22161

rashidakanchwala commented on src/actions/slice.js:
```diff
@@ -0,0 +1,28 @@
+  };
+};
+
+export const SLICE_PIPELINE = 'SLICE_PIPELINE';
```
nit, but maybe we can have consistent naming 

```
export const APPLY_SLICE = 'APPLY_SLICE';
export const SET_SLICE = 'SET_SLICE';
export const RESET_SLICE = 'RESET_SLICE';
```

Huongg replied:
hey good suggestion, I've updated to be consistent 

export const APPLY_SLICE_PIPELINE = 'APPLY_SLICE_PIPELINE';
export const SET_SLICE_PIPELINE  = 'SET_SLICE_PIPELINE ';
export const RESET_SLICE _PIPELINE = 'RESET_SLICE_PIPELINE ';

SajidAlamQB commented on src/components/node-list/node-list-tree-item.js:
```diff
@@ -13,10 +15,11 @@ const NodeListTreeItem = ({
-    className="pipeline-treeItem__root--overwrite"
-    key={data.id}
+    className={'pipeline-treeItem__root--overwrite'}
+    key={uniqueId(data.id)}
```
Could you explain what is happening here and why we need uniqueId?

SajidAlamQB commented on cypress/tests/ui/flowchart/flowchart.cy.js:
```diff
@@ -6,13 +6,14 @@ describe('Flowchart DAG', () => {
 
   beforeEach(() => {
     cy.enablePrettyNames(); // Enable pretty names using the custom command
+    cy.wait(500);
```
We are using a lot of `wait` is there not a way to check for some condition instead?

Huongg replied:
i can't think of anyway, i think because both of the setting modals and the hint modal would take less than 500 ms to disappear. that's why I put it there and wait until the modal is completely gone before trying to click on another item

Huongg replied:
and also i wonder why it doesnt fail anywhere else but only in this PR 🤔

SajidAlamQB reviewed (commented):
The slicing feature works great and looks awesome I've just left some minor comments.

Huongg commented on src/components/node-list/node-list-tree.js:
```diff
@@ -184,19 +211,25 @@ const TreeListProvider = ({
         onItemChange={onItemChange}
         onItemClick={onItemClick}
-        key={node.id}
+        key={uniqueId(node.id)}
```
we need this here because there's warning in the browser
<img width="1710" alt="Screenshot 2024-08-20 at 15 15 58" src="https://github.com/user-attachments/assets/97c35277-b322-4c59-92fc-c18190181303">

astrojuanlu commented:
Moved discussion to And the main discrepancy right now is that kedro run and kedro viz slice in opposite logic.
From node A to node C
In Kedro run, it's kedro run --from-nodes A --to-nodes C, that is
pipeline.from_nodes(A).to_nodes(C)
In kedro viz, it means Downstream of A until C and all upstream of C, that is
pipeline.from_nodes(A).to_nodes(C) | [pipeline.to](http://pipeline.to/)_nodes(C) <- The or operator is not supported in kedro run

Huongg commented:
hey @astrojuanlu thanks for reviewing the PR. You've spotted a very good bug there and as we discussed, I've addressed the command to only show `--to-output` or `--to-nodes`. Additionally, the missing command line is now included even when the node lacks a user-specified name.

- `kedro run --to-outputs`
<img width="703" alt="Screenshot 2024-08-23 at 17 21 06" src="https://github.com/user-attachments/assets/444521c9-92f5-4616-adaf-911739a0c96d">


-`kedro run --to-node`
<img width="1043" alt="Screenshot 2024-08-23 at 17 20 57" src="https://github.com/user-attachments/assets/47592ee4-81a6-466e-95bb-45dd897981e1">

jitu5 commented:
@Huongg @noklam @rashidakanchwala 
I tested the slicing feature in the VSCode extension. In terms of functionality, it is working fine, but there are some issues related to the UI.

On first click on node,  the Notification is not coming in center. 
<img width="1800" alt="Screenshot 2024-08-29 at 5 36 40 p m" src="https://github.com/user-attachments/assets/11ffa388-cf65-4dac-a5fb-a14f7bb5cd3f">

After slicing, the action bar at the bottom including run command and slicing button is hidden.
<img width="1800" alt="Screenshot 2024-08-29 at 5 37 14 p m" src="https://github.com/user-attachments/assets/120ec243-00f8-4422-877c-5b039104a28c">

If the same window is opened in full column layout, the entire action bar will be visible.
<img width="1800" alt="Screenshot 2024-08-29 at 5 38 18 p m" src="https://github.com/user-attachments/assets/cff224bf-68c8-4371-8b57-7836483e82bf">

Same for the sliced window.
<img width="1800" alt="Screenshot 2024-08-29 at 5 42 12 p m" src="https://github.com/user-attachments/assets/0ae33e97-2c23-4230-8a05-7357798a867b">
  
  
When the Kedro Viz window is opened for the first time, it appears in the second column layout and not in full screen. Considering this, we need to either fix the 'Notification on First Node Click' or disable the slicing feature for at least the `0.0.3` Kedro-VSCode release.

Huongg commented:
thanks @jitu5 I can see why this happened. At the moment the way I calculate the position of the box is based on the width  constant of sidebar, but i guess it's not the same value for the vscode. Good spot, and and thanks for checking this. Will have a think about how we should address this

stephkaiser commented:
Thanks @Huongg for addressing my comments! slicing is looking great and we are getting closer 

Some minor comments from me:

- btw the animation looks great!

- sometimes child nodes are still not highlighted at parent node group level (it only shows highlight at parent level if all child nodes in the group are selected)
<img width="1028" alt="Screenshot 2024-09-03 at 12 40 34" src="https://github.com/user-attachments/assets/52ecb123-b9f3-4067-94ac-1f4483d61a03">

- I wasn't able to test/see what the gradients look like when a run command is really long since it is much shorter now, maybe we can have a quick call about this

- after slicing, when clicking on a node in sliced view and metadata panel opens, the position of the run command bar is not in the center, but it is perfect when closing the left list panel:
<img width="1028" alt="Screenshot 2024-09-03 at 12 43 40" src="https://github.com/user-attachments/assets/76c5e1f9-8887-46da-bc8e-ff4b785df625">

- Question on changing the run command to only `--to-outputs`, is it very difficult to solve the run command issue? my concern is that the value and usefulness of this feature for users would now be less since the purpose of slicing is so they can run a specific part of the pipeline but if users don't have full control over that then users might not find enough value in using this feature. Wanted to hear your thoughts too @astrojuanlu

Huongg commented:
@stephkaiser yeah let's have a chat, I can show you the shadow box and also the context for the run command challenge we have.

Huongg commented:
hey @stephkaiser as discussed, I've addressed 2 points about the position of the run command, and the modular pipeline not highlighted when collapsed. Will create the ticket for the box shadow separately. Let me know if you're happy the changes or found some other issue 😄

stephkaiser commented:
it looks good to me! I have no more comments, thank you for all the hard work @Huongg!

SajidAlamQB reviewed (approved):
Awesome work @Huongg! 🔥

jitu5 reviewed (approved):
Thanks @Huongg. amazing work! 🔥🔥

ravi-kumar-pilla reviewed (approved):
Awesome work @Huongg and @stephkaiser !!

astrojuanlu reviewed (approved):
I was finally able to test this ❤️ Well done @Huongg!
